File: integrations/mcp/python/scripts/check_installed.py
"""Installed-package smoke against a host-launched local stdio MCP server."""
import asyncio
import json
import sys
from pathlib import Path

from mcp import ClientSession, StdioServerParameters, types
from mcp.client.stdio import stdio_client
from mcp.server import Server
from mcp.server.stdio import stdio_server
from purra.contracts import ExecutionState, ToolPolicy
from purra_mcp import McpCatalogMonitor, McpToolBinding, discover_mcp_tools


async def serve():
    server = Server('installed-local-fixture')
    @server.list_tools()
    async def tools():
        return [types.Tool(name='count', inputSchema={'type':'object','properties':{},'additionalProperties':False},
            outputSchema={'type':'object','properties':{'count':{'type':'integer'}},'required':['count'],'additionalProperties':False})]
    @server.call_tool()
    async def call(name, arguments):
        assert name == 'count' and arguments == {}
        return {'count':3}
    async with stdio_server() as streams:
        await server.run(*streams,server.create_initialization_options())


async def check():
    monitor = McpCatalogMonitor()
    params = StdioServerParameters(command=sys.executable,args=['-I',str(Path(__file__).resolve()),'serve'])
    try:
        async with stdio_client(params) as streams:
            async with ClientSession(*streams,message_handler=monitor.on_message) as client:
                initialized = await client.initialize()
                monitor.accept_protocol_version(initialized.protocolVersion)
                catalog = await discover_mcp_tools(client,'local',{'count':McpToolBinding('count',ToolPolicy(mode='read',title='Count'),None)},monitor=monitor)
                result = await catalog.registrations[0].handler(ExecutionState(),{})
                assert result.error_code is None
                assert json.loads(result.content)['structured'] == {'count':3}
                await client.send_ping()
        print('installed Python MCP stdio smoke passed')
    finally:
        monitor.close()


if __name__ == '__main__':
    asyncio.run(serve() if sys.argv[1:] == ['serve'] else check())

File: integrations/mcp/typescript/scripts/check-installed.mjs
// Installed-package smoke against a host-launched local stdio MCP server.
import assert from 'node:assert/strict';
import { fileURLToPath } from 'node:url';
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { ListToolsRequestSchema, CallToolRequestSchema, ToolListChangedNotificationSchema } from '@modelcontextprotocol/sdk/types.js';
import { discoverMcpTools, McpCatalogMonitor } from 'purra-mcp';

if (process.argv[2] === 'serve') {
  const server = new Server({name:'installed-local-fixture',version:'1'},{capabilities:{tools:{}}});
  server.setRequestHandler(ListToolsRequestSchema,async () => ({tools:[{name:'count',inputSchema:{type:'object',properties:{},additionalProperties:false},
    outputSchema:{type:'object',properties:{count:{type:'integer'}},required:['count'],additionalProperties:false}}]}));
  server.setRequestHandler(CallToolRequestSchema,async request => {
    assert.equal(request.params.name,'count');assert.deepEqual(request.params.arguments,{});
    return {content:[],structuredContent:{count:3}};
  });
  await server.connect(new StdioServerTransport());
} else {
  const monitor = new McpCatalogMonitor();
  const transport = new StdioClientTransport({command:process.execPath,args:[fileURLToPath(import.meta.url),'serve']});
  transport.setProtocolVersion = version => monitor.acceptProtocolVersion(version);
  const client = new Client({name:'consumer',version:'1'});
  client.setNotificationHandler(ToolListChangedNotificationSchema,n => monitor.onNotification(n));
  client.onclose = () => monitor.close();
  try {
    await client.connect(transport);
    const catalog = await discoverMcpTools(client,'local',{count:{localName:'count',policy:{mode:'read',title:'Count'},scope:null}},{monitor});
    const result = await catalog.registrations[0].run({}, {call:{id:'1',name:'count',arguments:{}}});
    assert.equal(result.errorCode,undefined);
    assert.deepEqual(result.content.structured,{count:3});
    await client.ping();
    console.log('installed TypeScript MCP stdio smoke passed');
  } finally { await client.close(); monitor.close(); }
}
